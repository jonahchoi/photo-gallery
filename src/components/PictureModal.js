import { deleteDoc, doc } from 'firebase/firestore';
import { deleteObject, ref } from 'firebase/storage';
import React, { useEffect, useState } from 'react'
import { Carousel, Dropdown, Image, Modal } from 'react-bootstrap'
import { appFirestore, appStorage } from '../firebase/config';

const PictureModal = ({ modalImgIdx, urlArray, showPictureModal, setShowPictureModal, setLoadSpinnerStatus }) => {

  const [ carouselIndex, setCarouselIndex ] = useState();

  useEffect(()=> {
    setCarouselIndex(modalImgIdx);
  }, [modalImgIdx]);

  let carouselArray = [];
  if(urlArray){
    for(let idx = 0; idx < urlArray.length; idx++){
      carouselArray.push(
        <Carousel.Item key={idx}>
          <Image 
            style={{width: '100%', maxHeight: '90vh'}} 
            src={urlArray[idx].url} 
          />
        </Carousel.Item>
      );
    }
  }

  const closePictureModal = () => {
    setShowPictureModal(false);
  }

  const deleteImage = (idx) => {

    setLoadSpinnerStatus('delete');
    const storageRef = ref(appStorage, urlArray[idx].filePath);
    const docInfo = doc(appFirestore, urlArray[idx].filePath);
    
    Promise.all([deleteObject(storageRef), deleteDoc(docInfo)])
      .then(()=> {
        console.log('Image deleted from storage and firestore.');
        setLoadSpinnerStatus('');
      })
      .catch((err) => {
        console.error(err);
        setLoadSpinnerStatus('');
      });

    closePictureModal();
  }

  const downloadImage = (idx) => {

    const xhr = new XMLHttpRequest();
    xhr.responseType = 'blob';
    xhr.onload = (event) => {
      const blob = xhr.response;
      const file = new File([blob], urlArray[idx].filePath, { type: blob.type });

      const downloadAnchor = document.createElement('a');
      downloadAnchor.setAttribute('href', URL.createObjectURL(file));
      downloadAnchor.setAttribute('download', file.name)
      
      document.body.appendChild(downloadAnchor);
      downloadAnchor.click();
      document.body.removeChild(downloadAnchor);
    }
    xhr.open('GET', urlArray[idx].url);
    xhr.send();

  }
 
  return (
    
    <Modal 
      dialogClassName='custom-modal-dialog' 
      show={showPictureModal} 
      onHide={closePictureModal} 
      centered 
      size='lg'
    >
      <Carousel 
        slide={false} 
        interval={null} 
        defaultActiveIndex={modalImgIdx}
        indicators={false}
        wrap={false}
        onSelect={(idx)=>setCarouselIndex(idx)}
      >
        {carouselArray}
      </Carousel>
      <Dropdown 
        className='modal-btn' 
        align='end'
        drop='up'
      >
        <Dropdown.Toggle>
          <i className="bi bi-three-dots-vertical" />
        </Dropdown.Toggle>
        <Dropdown.Menu>
          <Dropdown.Item 
            onClick={()=> downloadImage(carouselIndex)}
          >
              Download
          </Dropdown.Item>
          <Dropdown.Item 
            onClick={()=> deleteImage(carouselIndex)}
          >
            Delete
          </Dropdown.Item>
        </Dropdown.Menu>
      </Dropdown>
    </Modal>
  )
}

export default PictureModal
