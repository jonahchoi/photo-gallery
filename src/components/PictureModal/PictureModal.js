import { deleteDoc, doc } from 'firebase/firestore';
import { deleteObject, ref } from 'firebase/storage';
import React from 'react'
import { Carousel, Dropdown, Image, Modal } from 'react-bootstrap'
import { appFirestore, appStorage } from '../../firebase/config';

const PictureModal = ({ modalImgIdx, setModalImgIdx, urlArray, showPictureModal, setShowPictureModal, setLoadSpinnerStatus }) => {
  let carouselArray = [];
  for(let idx = 0; idx < urlArray?.length; idx++){
    carouselArray.push(
      <Carousel.Item key={idx}>
        <Image 
          style={{width: '100%', maxHeight: '90vh'}} 
          src={urlArray[idx].url} 
        />
      </Carousel.Item>
    );
  }

  const closePictureModal = () => {
    setShowPictureModal(false);
  }

  const deleteImage = (idx) => {

    setLoadSpinnerStatus('delete');
    const storageRef = ref(appStorage, urlArray[idx].filePath);
    const docInfo = doc(appFirestore, urlArray[idx].filePath);
    
    Promise.all([deleteObject(storageRef), deleteDoc(docInfo)])
      .then(()=> {
        console.log('Image deleted from storage and firestore.');
        setLoadSpinnerStatus('');
      })
      .catch((err) => {
        console.error(err);
        setLoadSpinnerStatus('');
      });

    closePictureModal();
  }

  const downloadImage = (idx) => {
    const xhr = new XMLHttpRequest();
    xhr.responseType = 'blob';
    xhr.onload = (event) => {
      const blob = xhr.response;
      const file = new File([blob], urlArray[idx].filePath, { type: blob.type });

      const downloadAnchor = document.createElement('a');
      downloadAnchor.setAttribute('href', URL.createObjectURL(file));
      downloadAnchor.setAttribute('download', file.name)
      
      document.body.appendChild(downloadAnchor);
      downloadAnchor.click();
      document.body.removeChild(downloadAnchor);
    }
    xhr.open('GET', urlArray[idx].url);
    xhr.send();

  }
 
  return (
    
    <Modal 
      dialogClassName='custom-modal-dialog' 
      show={showPictureModal} 
      onHide={closePictureModal} 
      centered 
      size='lg'
      data-testid='picture-modal'
    >
      <Carousel 
        slide={false} 
        interval={null} 
        defaultActiveIndex={modalImgIdx}
        indicators={false}
        wrap={false}
        onSelect={(idx)=>setModalImgIdx(idx)}
        data-testid='carousel'
      >
        {carouselArray}
      </Carousel>
      <Dropdown 
        className='modal-btn' 
        align='end'
        drop='up'
      >
        <Dropdown.Toggle id='drop-toggle' aria-label='dropdown toggle button'>
          <i className="bi bi-three-dots-vertical" />
        </Dropdown.Toggle>
        <Dropdown.Menu>
          <Dropdown.Item 
            onClick={()=> downloadImage(modalImgIdx)}
          >
              Download
          </Dropdown.Item>
          <Dropdown.Item 
            onClick={()=> deleteImage(modalImgIdx)}
          >
            Delete
          </Dropdown.Item>
        </Dropdown.Menu>
      </Dropdown>
    </Modal>
  )
}

export default PictureModal
