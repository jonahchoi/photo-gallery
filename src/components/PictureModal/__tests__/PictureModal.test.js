import React from 'react';
import { fireEvent, render } from '@testing-library/react';
import PictureModal from '../PictureModal';
import { act } from 'react-test-renderer';

const testUrlArray = [
  {url: 'https://cdn.pixabay.com/photo/2015/12/04/14/05/code-1076536_960_720.jpg',
  filepath: 'test.jpg'},
  {url: 'https://cdn.pixabay.com/photo/2016/11/19/14/00/code-1839406_960_720.jpg'}, 
  {url: 'https://cdn.pixabay.com/photo/2016/11/30/20/58/programming-1873854_960_720.png'}
];

const props = { 
  modalImgIdx: 0, 
  urlArray: testUrlArray, 
  showPictureModal: true
}

describe('PictureModal render test', () => {
  it('PictureModal renders', () => {
    const {getByTestId} = render(<PictureModal {...props} />);
    const pictureModal = getByTestId('picture-modal');
    expect(pictureModal).toBeTruthy();
  });
  
  it('PictureModal has correct number of images in carousel', () => {
    const { getByTestId, getAllByRole} = render(<PictureModal {...props} />);
    const pictures = getAllByRole('img');
    expect(pictures).toHaveLength(3);
    const pictureModal = getByTestId('picture-modal');
    expect(pictureModal).toMatchSnapshot();
  });
  
  it('Dropdown toggle works', async () => {
    
    const { getByRole, getByText } = render(<PictureModal {...props} />);
    
    await act( async ()=>{
      const toggleButton = getByRole('button', {name: 'dropdown toggle button'});
      await fireEvent.click(toggleButton);
    });
    const downloadButton = getByText('Download');
    const deleteButton = getByText('Delete');
    expect(downloadButton).toBeTruthy();
    expect(deleteButton).toBeTruthy();
    
  });
  
  it('Download button works', async () => {
    const mockDownload = jest.fn();
    
    const { getByRole, getByText } = render(<PictureModal {...props} />);
    
    await act( async ()=>{
      const toggleButton = getByRole('button', {name: 'dropdown toggle button'});
      await fireEvent.click(toggleButton);
    });
    const downloadButton = getByText('Download');
    downloadButton.onclick = mockDownload;
    await fireEvent.click(downloadButton);
    
    expect(mockDownload.mock.calls.length).toBe(1);
    
  });
  
  
})
