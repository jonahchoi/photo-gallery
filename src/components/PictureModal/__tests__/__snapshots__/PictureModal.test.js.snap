// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PictureModal render test PictureModal has correct number of images in carousel 1`] = `
<div
  class="modal-dialog custom-modal-dialog modal-lg modal-dialog-centered"
  data-testid="picture-modal"
>
  <div
    class="modal-content"
  >
    <div
      class="carousel"
      data-testid="carousel"
    >
      <div
        class="carousel-inner"
      >
        <div
          class="active carousel-item"
        >
          <img
            class=""
            src="https://cdn.pixabay.com/photo/2015/12/04/14/05/code-1076536_960_720.jpg"
            style="width: 100%; max-height: 90vh;"
          />
        </div>
        <div
          class="carousel-item"
        >
          <img
            class=""
            src="https://cdn.pixabay.com/photo/2016/11/19/14/00/code-1839406_960_720.jpg"
            style="width: 100%; max-height: 90vh;"
          />
        </div>
        <div
          class="carousel-item"
        >
          <img
            class=""
            src="https://cdn.pixabay.com/photo/2016/11/30/20/58/programming-1873854_960_720.png"
            style="width: 100%; max-height: 90vh;"
          />
        </div>
      </div>
      <a
        class="carousel-control-next"
        href="#"
        role="button"
        tabindex="0"
      >
        <span
          aria-hidden="true"
          class="carousel-control-next-icon"
        />
        <span
          class="visually-hidden"
        >
          Next
        </span>
      </a>
    </div>
    <div
      class="modal-btn dropup"
    >
      <button
        aria-expanded="false"
        aria-label="dropdown toggle button"
        class="dropdown-toggle btn btn-primary"
        id="drop-toggle"
        type="button"
      >
        <i
          class="bi bi-three-dots-vertical"
        />
      </button>
    </div>
  </div>
</div>
`;
